cac-agent/src/main/java/com/moesol/cac/agent/selector/AbstractSelectorKeyManager.java

```diff
@@ -234,17 +245,54 @@ private String maybeUseDefaultCertificateName(Principal[] issuers) {
 			if (ks == null) {
 				return new String[0];
 			}
+			Set<Principal> issuerSet = null;
+			if (issuers != null && issuers.length > 0) {
+				issuerSet = new HashSet<>(Arrays.asList(issuers));
+			}
 			ArrayList<String> asList = new ArrayList<String>();
 			Enumeration<String> aliases = ks.aliases();
 			while (aliases.hasMoreElements()) {
-				// TODO filter by keyType/issuers?
-				asList.add(aliases.nextElement());
+				// Ignore keyType filter to workaround bug JDK-8262186
+				String alias = aliases.nextElement();
+				if (ks.isKeyEntry(alias)) {
+					Certificate[] chain = ks.getCertificateChain(alias);
+					if (isX509Chain(chain)
+							&& (!checkCertIssuer || issuerMatches(chain, issuerSet))) {
+						asList.add(alias);
+					}
+				}
 			}
 			return asList.toArray(new String[asList.size()]);
 		} catch (KeyStoreException e) {
 			throw reportAndConvert(e);
 		}
 	}
+	private boolean isX509Chain(Certificate[] chain) {
+		if (chain == null || chain.length == 0) {
+			return false;
+		} else {
+			for (Certificate c : chain) {
+				if (c instanceof X509Certificate == false) {
+					return false;
+				}
+			}
+			return true;
+		}
+	}
+	private boolean issuerMatches(Certificate[] chain, Set<Principal> issuerSet) {
```

Comment:
Modeled after the logic in [OpenJDK](https://code.yawk.at/java/13/java.base/sun/security/ssl/X509KeyManagerImpl.java#772)